### ISSUE TITLE
Receiving error: Connection closed waiting for response -- version issue?

### ISSUE THREAD

**Issue by masonoise** [gocql/gocql#430]
I'm wondering if this could be a version mismatch error, though that seems unlikely -- perhaps I'm missing something. I'm using Cassandra 2.0.10, go version `go1.4.2 darwin/amd64`.

Connecting via cqlsh works:
````
$ CQLSH_PORT=9042 ./cqlsh
Connected to Test Cluster at localhost:9042.
[cqlsh 4.1.1 | Cassandra 2.0.10 | CQL spec 3.1.1 | Thrift protocol 19.39.0]
Use HELP for help.
cqlsh>
````

Connecting with gocql gives an error as follows:
````
2015/06/23 21:29:15 connect: failed to connect to "localhost:9042": gocql: connection closed waiting for response
runtime error: invalid memory address or nil pointer dereference
````

With the code:
````
cluster := gocql.NewCluster("localhost")
cluster.Keyspace = "testing"
session, _ := cluster.CreateSession()
defer session.Close()
err := session.Query("SELECT summary,time from testing where foo = 'bar'").Scan(&summary, &time)
````

I must be missing something, but everything matches up so it seems like it should be working. Any thoughts would be great, I'm happy to test things out if it'll help. Thanks very much.

**Comment by Zariel** [comment]
Thanks for the report. Could you specify what version of gocql you are using, and the stack trace from the nil pointer. Also it looks like you are missing an error being returned from createSession.

**Comment by masonoise** [comment]
Sure, thanks for checking. Here's the full stack trace. I'm very new to go so I'm not sure how to check the version of gocql, but I just pulled it down yesterday so it's presumably the latest. The commit sha in the gocql dir is `46cf8d711076fd4e4bdd89f680b127a9b46ed90a`. I will investigate the createSession call and see if I can find any error that's not showing up there.

````
2015/06/23 21:29:15 connect: failed to connect to "localhost:9042": gocql: connection closed waiting for response
runtime error: invalid memory address or nil pointer dereference
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/recover.go:41 (0x46558)
        func.019: trace := debug.Stack()
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/asm_amd64.s:401 (0x3cbf5)
        call16: CALLFN(·call16, 16)
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/panic.go:387 (0x14b28)
        gopanic: reflectcall(unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/panic.go:42 (0x13e4e)
        panicmem: panic(memoryError)
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/sigpanic_unix.go:26 (0x1a3c0)
        sigpanic: panicmem()
/Users/mason/workspace/gowork/src/github.com/gocql/gocql/session.go:181 (0xb1cf8)
        (*Session).Close: s.closeMu.Lock()
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/asm_amd64.s:401 (0x3cbf5)
        call16: CALLFN(·call16, 16)
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/panic.go:387 (0x14b28)
        gopanic: reflectcall(unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/panic.go:42 (0x13e4e)
        panicmem: panic(memoryError)
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/sigpanic_unix.go:26 (0x1a3c0)
        sigpanic: panicmem()
/Users/mason/workspace/gowork/src/github.com/gocql/gocql/session.go:146 (0xb19b8)
        (*Session).Query: s.mu.RLock()
/Users/mason/workspace/gowork/src/github.com/masonoise/json-api/api.go:26 (0x26fc)
        func.001: err := session.Query("SELECT summary,time from testing where zipcode = '94110'").Scan(&summary, &time)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/router.go:56 (0x46947)
        func.021: handler(writer, request)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/content_type_checker.go:38 (0x45cfc)
        func.010: handler(w, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/json_indent.go:35 (0x45ff4)
        func.014: handler(writer, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/recover.go:57 (0x46807)
        func.020: h(w, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/powered_by.go:26 (0x4629c)
        func.017: h(w, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/recorder.go:22 (0x4637e)
        func.018: h(writer, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/timer.go:20 (0x46a89)
        func.023: h(w, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/access_log_apache.go:87 (0x45625)
        func.001: h(w, r)
/Users/mason/workspace/gowork/src/github.com/ant0ine/go-json-rest/rest/middleware.go:70 (0x461a0)
        func.016: handler(writer, request)
/usr/local/Cellar/go/1.4.2/libexec/src/net/http/server.go:1265 (0x5eab1)
        HandlerFunc.ServeHTTP: f(w, r)
/usr/local/Cellar/go/1.4.2/libexec/src/net/http/server.go:1703 (0x60b3a)
        serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
/usr/local/Cellar/go/1.4.2/libexec/src/net/http/server.go:1204 (0x5e607)
        (*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
/usr/local/Cellar/go/1.4.2/libexec/src/runtime/asm_amd64.s:2232 (0x3ed11)
````

**Comment by Zariel** [comment]
As there is no mention of gocql in the stack trace it looks as if it is failing to connect so there will be an error returned form `CreateSession` which then means when accessing the `Query` method it returns a nil iter which then calling `exec` I presume causes a nil pointer.

**Comment by Zariel** [comment]
To clarify, you are getting a panic because the returned session is `nil` because `err != nil`, then in the defer `db.Close()` it causes a nil pointer dereference when `Close` dereferences `s`.

Instead try this,

```go
cluster := gocql.NewCluster("localhost")
cluster.Keyspace = "testing"
session, err := cluster.CreateSession()
if err != nil {
    // or otherwise handle error
    log.Println(err)
    return
}
defer session.Close()
err = session.Query("SELECT summary,time from testing where foo = 'bar'").Scan(&summary, &time)
```

**Comment by masonoise** [comment]
Thanks for the feedback. I made some updates but it's still failing to connect, even though I can connect using cqlsh to the same host/port, and I'm still not sure why. Just to make things super clear I updated the code:

````
cluster := gocql.NewCluster("localhost")
cluster.Keyspace = "testing"
session, create_err := cluster.CreateSession()
if create_err != nil {
    println("Error creating the session")
    log.Fatal(create_err)
}
println("Created session, getting ready to query")
defer session.Close()
query_err := session.Query("SELECT summary,time from testing where zipcode = '94110'").Scan(&summary, &time)
if query_err != nil {
    println("Error querying")
    log.Fatal(query_err)
}
````

With the result:

````
2015/06/24 22:52:11 connect: failed to connect to "localhost:9042": gocql: connection closed waiting for response
Error creating the session
2015/06/24 22:52:11 no connections were made when creating the session
exit status 1
````

Can you think of any ways to get more details about why it's not able to create the session? It's just very strange that I can connect from command-line, but not from the code, and I feel like I must be missing something obvious here... Thanks again for your help.

**Comment by nemothekid** [comment]
The `2015/06/24 22:52:11 connect: failed to connect to "localhost:9042": gocql: connection closed waiting for response` bit is reported from [inside gocql](https://github.com/gocql/gocql/blob/677750eba8a8db8e2b434e1dda050282b95bf491/connectionpool.go#L230), so despite the panic, there is definitely an issue connecting to the cluster.

The error `connection closed waiting for response` is an internal gocql error thats only triggered when a connection is closed, so there is some error happening before then caused by your setup (or a bug in the code). In any case, it looks like the negotiating the `startup frame` (or the `authentication frame`) is failing, and the library is eating the exact error.

If you are on 46cf8d711076fd4e4bdd89f680b127a9b46ed90a, you can try adding the following to [conn.go](https://github.com/gocql/gocql/blob/master/conn.go#L624), which may help in deciphering what is going on.

```diff
...
func (c *Conn) closeWithError(err error) {
	if !atomic.CompareAndSwapInt32(&c.closed, 0, 1) {
		return
	}
+   log.Printf("Closing connection with error: %v\n", err)
	close(c.quit)
...
```

**Comment by masonoise** [comment]
Thanks for the help, everyone. I tested further and with that debugging in conn.go I saw only "Closing connection with error: EOF". So then I pointed it to our staging Cassandra cluster, which has a newer version of Cassandra, and it works: `[cqlsh 4.1.1 | Cassandra 2.1.5 | CQL spec 3.1.1 | Thrift protocol 19.39.0]`. It seems possible that there's some incompatibility here with Cassandra 2.0.10, which is what I'm running on my laptop. I'll try upgrading it and see if the problem goes away, and will report back. Thanks again.
